### PR TITLE
Bug 1438205 - Preserve comments in progress across page reloads

### PR DESCRIPTION
**Pull request by purelogiq**
### How it works
- When you input (type or copy paste) into the comment box, we save the text of the box into localStorage.
- When the bug page loads, or tab is switched to (if 2 tabs with the same bug are open) we populate the comment box with the saved text, if it exists.
- When you save a bug we clear the localStorage.
- This solves the problem mentioned in https://bugzilla.mozilla.org/show_bug.cgi?id=1438205 and just in general allows people to start typing a comment, close the page, and come back to it tomorrow.

### CODE REVIEW

**Comment by purelogiq** [comment]
### Questions:

- In general for UI bmo work, we don't have to worry about super old IE browsers right? For browsers that we expect to be running the JS, can we assume >IE10 or something?

- Is this file, `bug_modal.js` the right place to put this stuff? It seems to be where all the other JS for the show bug page lives.

- How do you want to handle comment expiration, if at all? If the user hasn't submitted a saved comment in like 30 days, should we just remove it or leave it saved forever?

- Overall feedback appreciated :)

**Comment by dylanwh** [comment]
### Questions:

- Q: _In general for UI bmo work, we don't have to worry about super old IE browsers right? For browsers that we expect to be running the JS, can we assume >IE10 or something?_
  - A: This is covered here: https://wiki.mozilla.org/BMO#Browser_Support_For_BMO

- Q: _Is this file, `bug_modal.js` the right place to put this stuff? It seems to be where all the other JS for the show bug page lives._
  - A: __Yes. bug_modal.js, and for styles bug_modal.css__
  - This is all very messy right now, I'll develop a plan for making this cleaner.

- Q: _How do you want to handle comment expiration, if at all? If the user hasn't submitted a saved comment in like 30 days, should we just remove it or leave it saved forever?_
  - A: >7 days, expire.
  - Also, if possible remove old saved comments for unvisited bugs too.

**Comment by purelogiq** [comment]
## Updated:

- Refactored the core functions to be vanilla ES6. Only the event handlers near the bottom are left with jQuery style to be consistent in that part of the file.

- Added expiration for saved comments older than a week. This happens once on page load and will clear out all comments that have expired.

- Added some safe guards around saving 0-length or comments > 65535 characters. 

- Tested with large comments. **Result:** Pasting a comment that is 65535 characters causes about a quarter-second jitter, but, afterwards editing the large comment is smooth as usual.

- If the user has the autosize option enabled, we respect that and resize when updating the textbox from localstorage.

@dylanwh FYI autosize seems to be really wonky with large comments (it jitters between the top and bottom of the textbox when typing), even before this patch and on production. Since the vagrant vm has it enabled by default but production doesn't, for a long time I thought it was because of my changes. On my vm I've made disabled 'Expand the comment box dynamically' and enabled 'Zoom textareas large when in use' in General Preferences - do you want to make that the default for the vagrant vm?

**Comment by dylanwh** [comment]
> do you want to make that the default for the vagrant vm?
Yes, and perhaps that code needs to be removed in a future bug?

File: extensions/BugModal/web/bug_modal.js [hunk 1]
```diff
@@ -96,6 +96,54 @@ $(function() {
...
+            }
+        }
+    }
+
+    function expireSavedComments() {
+        const AGE_THRESHOLD = 7 * 24 * 60 * 60 * 1000; // 7 days in milliseconds.
+        let expiredKeys = [];
+        for(let i = 0; i < localStorage.length; i++) {
+            let key = localStorage.key(i);
+            if(key.match(/^bug-modal-saved-comment-/)) {
```
Comment (dylanwh) [inline extensions/BugModal/web/bug_modal.js]: space here

File: extensions/BugModal/web/bug_modal.js [hunk 2]
```diff
@@ -96,6 +96,54 @@ $(function() {
...
+            if (BUGZILLA.user.settings.autosize_comments) {
+                autosize.update(commentBox);
+            }
+        }
+    }
+
+    function expireSavedComments() {
+        const AGE_THRESHOLD = 7 * 24 * 60 * 60 * 1000; // 7 days in milliseconds.
+        let expiredKeys = [];
+        for(let i = 0; i < localStorage.length; i++) {
```
Comment (dylanwh) [inline extensions/BugModal/web/bug_modal.js]: space here

File: extensions/BugModal/web/bug_modal.js [hunk 3]
```diff
@@ -96,6 +96,54 @@ $(function() {
...
+    function clearSavedBugComment() {
+        let key = `bug-modal-saved-comment-${BUGZILLA.bug_id}`;
+        localStorage.removeItem(key);
+    }
+
+    function restoreSavedBugComment() {
+        expireSavedComments();
+        let key = `bug-modal-saved-comment-${BUGZILLA.bug_id}`;
+        let value = JSON.parse(localStorage.getItem(key));
+        if(value){
```
Comment (dylanwh) [inline extensions/BugModal/web/bug_modal.js]: space here

File: extensions/BugModal/web/bug_modal.js [hunk 4]
```diff
@@ -96,6 +96,54 @@ $(function() {
         $('#editing').val('');
     }
 
+    function saveBugComment(text) {
+        if(text.length < 1) return clearSavedBugComment();
```
Comment (dylanwh) [inline extensions/BugModal/web/bug_modal.js]: space here and the next line

**Pull request merged**